scheduler/schedule.go

```diff
@@ -0,0 +1,93 @@
+package scheduler
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/axsh/openvdc/handlers"
+	"github.com/axsh/openvdc/model"
+)
+
+var (
+	// Time to wait for the offer to arrive
+	waitOfferSec = 5
+
+	instanceSchedulerHandlers = make(map[string]handlers.InstanceScheduleHandler)
+	schedule                  *Schedule
+)
+
+func init() {
+	schedule = newSchedule()
+}
+
+func RegisterInstanceScheduleHandler(name string, i handlers.InstanceScheduleHandler) error {
+	if _, exists := instanceSchedulerHandlers[name]; exists {
+		return fmt.Errorf("Duplicated name for instance schedule handler: %s", name)
+	}
+	instanceSchedulerHandlers[name] = i
+	return nil
+}
+
+type Schedule struct {
+	*sync.Mutex
+	storedOffers map[string]*model.VDCOffer
+}
+
+func newSchedule() *Schedule {
+	return &Schedule{
+		Mutex:        new(sync.Mutex),
+		storedOffers: make(map[string]*model.VDCOffer),
+	}
+}
+
+// TODO write definition of openvdc offer (interface | struct)
+func (s *Schedule) StoreOffer(offer *model.VDCOffer) {
+	schedule.Lock()
+	defer schedule.Unlock()
+	schedule.storedOffers[offer.SlaveID] = offer
+}
+
+func (s *Schedule) Assign(inst *model.Instance) error {
```

Comment:
Missing to take mutex lock.

Reply:
Added RLock/RUnlock next to schedule.storedOffers's "for range" function.